spec: Light client attack detector (#164)
tendermint/spec: rust-spec/lightclient/detection/detection.md

`VerifyToTarget` 
being a lightblock; new version number of function tag;

- We should clarify what is the expectation of VerifyToTarget
so if it returns TimeoutError it can be assumed faulty. I guess that
VerifyToTarget with correct full node should never terminate with
TimeoutError.

- We need to introduce a new version number for the new
specification. So we should decide how
  to handle that.

# Light Client Attack Detector

In this specification, we strengthen the light client to be resistant
against so-called light client attacks. In a light client attack, all
the correct Tendermint full nodes agree on the sequence of generated
blocks (no fork), but a set of faulty full nodes attack a light client
by generating (signing) a block that deviates from the block of the
same height on the blockchain. In order to do so, some of these faulty
full nodes must have been validators before and violate
[[TMBC-FM-2THIRDS]](TMBC-FM-2THIRDS-link), as otherwise, if
[[TMBC-FM-2THIRDS]](TMBC-FM-2THIRDS-link) would hold,
[verification](verification) would satisfy
[[LCV-SEQ-SAFE.1]](LCV-SEQ-SAFE-link).

An attack detector (or detector for short) is a mechanism that is used

> by the light client [supervisor](supervisor) after

Reviewer: Supervisor link is broken.

Author: We don't have the files on master yet. I will fix that then.

[verification](verification) of a new light block
with the primary, to cross-check the newly learned light block with
other peers (secondaries).  It expects as input a light block with some
height *root* (that serves as a root of trust), and a verification
trace (a sequence of lightblocks) that the primary provided.

In case the detector observes a light client attack, it computes
evidence data that can be used by Tendermint full nodes to isolate a
set of faulty full nodes that are still within the unbonding period

> (more than 1/3 of the validator set of some block on the chain),

Reviewer: ```suggestion
(more than 1/3 of the voting power of the validator set at some block of the chain),
```

and report them via ABCI to the application of a Tendermint blockchain
in order to punish faulty nodes.

## Context of this document

The light client [verification](verification) specification is
designed for the Tendermint failure model (1/3 assumption)
[[TMBC-FM-2THIRDS]](TMBC-FM-2THIRDS-link).  It is safe under this
assumption, and live if it can reliably (that is, no message loss, no
duplication, and eventually delivered) and timely communicate with a

> correct full node. If this assumption is violated, the light client

Reviewer: ```suggestion
correct full node. If [[TMBC-FM-2THIRDS]](TMBC-FM-2THIRDS-link) assumption is violated, the light client
```

The light client maintains a simple address book containing addresses
of full nodes that it can pick as primary and secondaries.  To obtain
a new light block, the light client first does
[verification](verification) with the primary, and then cross-checks
the light block (and the trace of light blocks that led to it) with
the secondaries using this specification.

## Tendermint Consensus and Light Client Attacks

In this section we will give some mathematical definitions of what we
mean by light client attacks (that are considered in this
specification) and how they differ from main-chain forks. To this end
we start by defining some properties of the sequence of blocks that is
decided upon by Tendermint consensus in normal operation (if the
Tendermint failure model holds
[[TMBC-FM-2THIRDS]](TMBC-FM-2THIRDS-link)),
and then define different
deviations that correspond to attack scenarios.

#### **[TMBC-GENESIS.1]**

Let *Genesis* be the agreed-upon initial block (file).

#### **[TMBC-FUNC-SIGN.1]**

Let *b* and *c* be two light blocks with *b.Header.Height + 1 =
c.Header.Height*. We define the predicate **signs(b,c)** to hold

> iff *c.Header.LastCommit* is in *PossibleCommit(b)*.

Reviewer 2: Is this the same as the two conditions b.NextValidatorsHash = c.ValidatorsHash && 2/3 of these Validators signed c?

Author: Yes.  I will add a comment to make this clear.

Author: added

[[TMBC-SOUND-DISTR-POSS-COMMIT.1]](TMBC-SOUND-DISTR-POSS-COMMIT-link).
 
#### **[TMBC-FUNC-SUPPORT.1]**

Let *b* and *c* be two light blocks. We define the predicate
**supports(b,c,t)** to hold iff

- *t - trustingPeriod < b.Header.Time < t*
- the voting power in *b.NextValidators* of nodes in *c.Commit*
  is more than 1/3 of *TotalVotingPower(b.Header.NextValidators)*

> That is, if the [Tendermint failure model](TMBC-FM-2THIRDS-link)

> > holds, then *c* has been signed by at least on correct full node, cf.

Reviewer: ```suggestion
> holds, then *c* has been signed by at least one correct full node, cf.
```

> > holds, then *c* has been signed by at least on correct full node, cf.

Reviewer: Do we need to add an additional assumptions that b comes from Tendermint chain?

Author: No. I also will use that to define a verification trace which does not necessarily mean that the b is from the chain. This will come later when `supports` is used in other definitions.

> The following captures a light client fork. There is no fork up to
> the height of block b. However, c is of that height, is different,
> and passes skipping verification. It is a stricter property than
> [[TMBC-LC-ATTACK.1]](#TMBC-LC-ATTACK1), as
> [[TMBC-LC-ATTACK.1]](#TMBC-LC-ATTACK1) requires that no correct full
> node is affected.

#### **[TMBC-BLOCKS-LCFORK.1]**

Let *a*, *b*, *c*, be light blocks and *t* a time. We define
*light-client-fork(a,b,c,t)* iff

- *sign-skip-match(a,b,c,t) = false* and
- *sequ-rooted(b)* and
- *b* is "unique", that is, for all *d*,  *sequ-rooted(d)* and
     *d.Header.Height = b.Header.Height* implies *d = b*

> Finally, let us also define bogus blocks that have no support.
> Observe that bogus is even defined if there is a fork.
> Also, for the definition it would be sufficient to restrict *a* to
> *a.height < b.height* (which is implied by the definitions which
> unfold until *supports()*).

#### **[TMBC-BOGUS.1]**

Let *b* be a light block and *t* a time. We define *bogus(b,t)* iff

- *sequ-rooted(b) = false* and
- for all *a*, *sequ-rooted(a)* implies *skip-trace(a,b,t) = false*
  
### Informal Problem statement

> There is no sequential specification: the detector only makes sense

Reviewer 3: There is a sequential specification using a global state, that contains a set of light blocks. You can use sing and support to define input output relations, i.e., a list, resp. a tree (see our discussion):).

- *a(h) = c* and
- *supports( a(i), a(i+1), t)*, for all i, *1 <= i < h*.

> We call such a sequence *a(1)*, ... *a(h)* a **verification trace**.

Reviewer: Why skipping condition is called skip-root?

Author: For historic reasons ;-). 
I guess `skip-trace` would be a better name. I will update that.

> The following formalizes that two light blocks of the same height
> should agree on the content of the header. Observe that *b* and *c*
> may disagree on the Commit. This is a special case if the canonical
> commit has not been decided on, that is, if b.Header.Height is the
> maximum height of all blocks decided upon by Tendermint at this
> moment.

#### **[TMBC-SIGN-SKIP-MATCH.1]**

Let *a*, *b*, *c*, be light blocks and *t* a time, we define
*sign-skip-match(a,b,c,t) = true* iff the following implication
evaluates to true:

- *sequ-rooted(a)* and
- *b.Header.Height = c.Header.Height* and
- *skip-root(a,b,t)*
- *skip-root(a,c,t)*

implies *b.Header = c.Header*.

> Observe that *sign-skip-match* is defined via an implication. If it
> evaluates to false this means that the left-hand-side of the
> implication evaluates to true, and the right-hand-side evaluates to
> false. In particular, there are two **different** headers *b* and
> *c* that both can be verified from a common block *a* from the

> > chain. Thus, thee following describes an attack.

Reviewer: ```suggestion
> chain. Thus, the following describes an attack.
```

- there is a peer p1 that has a sequence *chain* of blocks from *a* to *b*
- *skip-root(a,c,t)*: by [[TMBC-SKIP-ROOT.1]](#TMBC-SKIP-ROOT1) there is a
  verification trace *v* of the form *a = v(1)*, ... *v(h) = c*

Evidence for p1 (that proves an attack) consists for index i
of v(i) and v(i+1) such that

- E1(i). v(i) is equal to the block of *chain* at height v(i).Height, and
- E2(i). v(i+1) that is different from  the block of *chain* at
  height v(i+1).height

> Observe p1 can
>
> - check that v(i+1)  differs from its block at that height, and
> - verify v(i+1) in one step from v(i) as v is a verification trace.

**Proposition.** In the case of attack, evidence exists.  
*Proof.* First observe that

- (A). (NOT E2(i)) implies E1(i+1)

Now by contradiction assume there is no evidence. Thus

- for all i, we have NOT E1(i) or NOT E2(i)
- for i = 1 we have E1(1) and thus NOT E2(1)
  thus by induction on i, by (A) we have for all i that **E1(i)**
- from attack we have E2(h-1), and as there is no evidence for
  i = h - 1 we get **NOT E1(h-1)**. Contradiction.
QED.

#### **[TMBC-LC-EVIDENCE-DATA.1]**

To prove the attack to p1, because of Point E1, it is sufficient to
submit

> - v(i).Height (rather than v(i)).

Reviewer: Callum made an observation that in case of Equivocation and Amnesia conflicting blocks are for the same height and they have the same validator set. So we don't need root height, i.e., we could set root height to v(i+1).Height. But maybe this can be seen as an optimisation (I think this is how it is being implemented in Go version).

Author: I don't see what the "optimization" here would be. From a semantic viewpoint, currently the evidence says:  
"Some of the **next validators** in v(i).Height messed up v(i+1)"

If we would change that the semantics it would say:
"Some of the next validators in v(i).Height messed up v(i+1), unless v(i).Height = v(i+1).Height in which case the **validators** in v(i).Height messed up v(i+1)"

My feeling is that it is more natural to treat the special case "v(i).Height + 1= p v(i+1).Height" on the receiving side of the evidence rather on the sending side, and check this condition there.

Also, we do not check at the light client what kind of attack we have  (we decided to not add this information in the evidence). So it is not clear how to figure out that case.

@cmwaters : do I miss something here?

the protocols when communicating with the light client, this
specification also considers the generation of evidence that will
also be processed by the Tendermint blockchain.

#### **[LCD-IP-MODEL.1]**

The detector is designed under the assumption that

- [[TMBC-FM-2THIRDS]](TMBC-FM-2THIRDS-link) may be violated
- there is no fork on the main chain.

> As a result some faulty full nodes may launch an attack on a light
> client. 

The following requirements are operational in that they describe how
things should be done, rather than what should be done. However, they
do not constitute temporal logic verification conditions. For those,
see [LCD-DIST-*] below.

The detector is called in the [supervisor](supervisor) as follows

```go
Evidences := AttackDetector(root_of_trust, verifiedLS);`
```

where

- `root-of-trust` is a light block that is trusted (that is,
except upon initialization, the primary and the secondaries
agreed on in the past), and
- `verifiedLS` is a lightstore that contains a verification trace that
  starts from a lightblock that can be verified with the
  `root-of-trust` in one step and ends with a lightblock of the height
  requested by the user

> - `Evidences` is a list of evidence for misbehavior

Reviewer: ```suggestion
- `Evidences` is a list of evidences for misbehavior
```

   
        EvidenceForSecondary, newroot, secondary_trace, result :=
                CreateEvidenceForPeer(secondary, 
				                      root, 
									  primary_trace);
        if result == FaultyPeer {
            Replace_Secondary(root);
        }
        else if result == FoundEvidence {
            // the conflict is not bogus
            Evidences.Add(EvidenceForSecondary);
            // we replay the secondary trace with the primary, ...
            EvidenceForPrimary, _, result := 
				    CreateEvidenceForPeer(primary,
                                          newroot,
                                          secondary_trace);
            if result == FoundEvidence {
                Evidences.Add(EvidenceForPrimary);
            }
            // At this point we do not care about the other error
            // codes. We already have generated evidence for an
            // attack and need to stop the lightclient. It does not
            // help to call replace_primary. Also we will use the
            // same primary to check with other secondaries in
            // later iterations of the loop
         }

>          else {

Reviewer: I don't get what this branch is doing.

Author: Sorry. I changed the control flow last second. I will fix that right now. We do nothing here. This is the case where the secondary agrees with the primary.

Author: fixed

>          else {
>              // first the secondary reported a faulty block but
>              // then reported a matching block for the same height
>              // first block was most likely bogus
>              Replace_Secondary(root);
>          }

Author: ```suggestion
         // In the case where the secondary reports NoEvidence
         // we do nothing
```

    }
    return Evidences; 
}
```

- Expected precondition

>     - root and primary trace satisfy are a verification trace

Reviewer: ```suggestion
    - root and primary trace are a verification trace
```

                                          secondary_trace);
            if result == FoundEvidence {
                Evidences.Add(EvidenceForPrimary);
            }
            // At this point we do not care about the other error
            // codes. We already have generated evidence for an
            // attack and need to stop the lightclient. It does not
            // help to call replace_primary. Also we will use the
            // same primary to check with other secondaries in
            // later iterations of the loop
         }
         // In the case where the secondary reports NoEvidence
         // we do nothing
    }
    return Evidences; 
}
```

- Expected precondition
    - root and primary trace are a verification trace
- Expected postcondition
    - solves the problem statement (if attack found, then evidence is reported)
- Error condition
    - `ErrorTrustExpired`: fails if root expires (outside trusting
    period) [[LCV-INV-TP.1]](LCV-INV-TP1-link)
    - `ErrorNoPeers`: if no peers are left to replace secondaries, and
    no evidence was found before that happened

----

```go
func CreateEvidenceForPeer(peer PeerID, root LightBlock, trace LightStore)

>                           (Evidence, LightStore, result) {

Reviewer: I think we are missing one return value as it returns 4 values.

Author: yes, I added the LightBlock at the second position

    common := root;

    for i in 1 .. len(trace) {
        auxLS, result := VerifyToTarget(peer, common, trace[i].Header.Height)
  
        if result != ResultSuccess {
            // something went wrong; peer did not provide a verifyable block
            return (nil, nil, nil, FaultyPeer)
        }
        else {
            if auxLS.LatestVerified().Header != trace[i].Header {
                // the header reported by the peer differs from the
                // reference header in trace but both could be
                // verified from common in one step.
                // we can create evidence for submission to the secondary
                ev := new InternalEvidence;
                ev.Evidence.ConflictingBlock := trace[i];
                ev.Evidence.CommonHeight := common.Height;
                ev.Peer := peer
                return (ev, common, auxLS, FoundEvidence)
            }
            else {
                // the peer agrees with the trace, we move common forward
                // we could delete auxLS as it will be overwritten in
                // the next iteration

>                 common := trace[i].Header

Reviewer: ```suggestion
                common := trace[i]
```

Reviewer: I think we are passing LightBlock to VerifyToTarge, not header.

            }
        }
    }
    return (nil, nil, nil, NoEvidence)
}
```
- Expected precondition
    - root and trace are a verification trace
- Expected postcondition
    - finds evidence where trace and peer diverge
- Error condition

>     - `ErrorTrustExpired`: fails if root expires (outside trusting

Reviewer: Should we return error obtain from VerifyToTarget so this is true?

Author: I added a clarification
